### ISSUE TITLE
Heap corruption with MSVC debug assertions on

### ISSUE THREAD

**Issue by AnyOldName3** [openscenegraph/OpenSceneGraph#985]
I've built the OpenSceneGraph-3.6 branch with MSVC 2017 with the Debug configuration, which CMake enables a lot of assertions for, like STL range checks and certain kinds of heap corruption.

I've got a TGA file that I'm unconvinced is being read correctly, so I'm trying to run
```
osgviewer --image "path/to/file.tga" -O "ignoreTga2Fields"
```

I've not got particular far with that, though. I've seen two failed assertions so far (obviously on different runs, as the application dies when the assertions fail):
* Here, when the TGA plugin registers its extensions and options: https://github.com/openscenegraph/OpenSceneGraph/blob/OpenSceneGraph-3.6/src/osgPlugins/tga/ReaderWriterTGA.cpp#L808. The heap corruption was in the `new` operator when allocating a string.
* Here, when the TGA plugin deleted itself after the `ref_ptr` in its `osgDB::RegisterReaderWriterProxy` went out of scope during destruction: https://github.com/openscenegraph/OpenSceneGraph/blob/OpenSceneGraph-3.6/src/osg/Referenced.cpp#L292

I don't think it's especially likely that the TGA plugin itself is doing something to cause heap corruption, as I had these assertions enabled when I was last working on it, so I guess maybe something's unhappy in `osgDB`.

When it *does* get past creation of the plugin and the error manifests during exit, it's also printing an unexpected 
```
Error reading file C:\Games\Morrowind\Mods\Fireflies Invade Morrowind\Textures\glowarea33.tga: read error (Could not find plugin to read objects from file "C:\Games\Morrowind\Mods\Fireflies Invade Morrowind\Textures\glowarea33.tga".)
C:\C++Libraries\OpenSceneGraph\OpenSceneGraph\bin\osgviewerd.exe: No data loaded
```
message, too. Obviously, the TGA plugin *should* be present.

I'll update this as I learn more.

**Comment by AnyOldName3** [comment]
The assertions don't trigger with RelWithDebInfo, but the same read error occurs. I've had a cursory look through the history for `osgDB` and don't see anything that I think is likely to have broken image plugin registration.

I also get similar problems with PNG images and the PNG reader if I specify a path to a PNG image, so either the entirety of image loading is dead, which I feel someone would have noticed, or the `--image` mode of the `osgviewer` example is dead.

**Comment by AnyOldName3** [comment]
The OpenSceneGraph-3.6.5 tag works, so I'm going to bisect.

**Comment by AnyOldName3** [comment]
Bisection shows that https://github.com/openscenegraph/OpenSceneGraph/commit/c3108697dea79c188e5869fb04eb46e2c399cde5#diff-324307a67f928d09364e2901fb7c64c3 is the commit that caused the regression. I initially dismissed that as not looking like it would break anything, so I'll try and track down why it's breaking things.

**Comment by AnyOldName3** [comment]
I think I've figured it out, and it's a build system problem rather than a code problem per-se. In the past, I've manually set up the osgviewer application to depend on the image plugins so they'd get rebuilt whenever it did. That's been lost at some point, so the plugins were left as they'd been when I last built them. The linked commit changed the ABI, so obviously those plugins weren't compatible anymore, but as the plugin directory name is dependent on the OSG release version number rather than the ABI version number, they were still getting loaded.

I'm going to close this now, as the heap corruption isn't a bug in OSG that needs to be fixed, but maybe it's a good idea to take a look at making it so plugins with a mismatched ABI version can't be loaded.
